r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -131,6 +132,56 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
         Ok(res)
     }
 
+    /// Construct a NonNativeFieldMulResult gadget from a NonNativeFieldGadget
+    pub(crate) fn from_non_native_field_gadget<CS: ConstraintSystemAbstract<ConstraintF>>(mut cs: CS, input_gagdet: &NonNativeFieldGadget<SimulationF, ConstraintF>,
+        params: &NonNativeFieldParams) -> Self {
+
+        debug_assert!(
+            input_gagdet.check(),
+            "convert to mul result gagdet from non native gadget: check() failed on input gadget"
+        );
+
+        let mut limbs = vec![FpGadget::<ConstraintF>::from_value(&mut cs, &ConstraintF::zero()); params.num_limbs-1];
+        limbs.extend_from_slice(&input_gagdet.limbs[..]);
+        // set num_add to ceil(input_gadget.num_add/2^bits_per_limb) as a limb of
+        // NonNativeFieldMulResultGadget has 2*bits_per_limb for each limb
+        let num_add_over_normal_form = &input_gagdet.num_of_additions_over_normal_form/BigUint::from(2usize).pow(params.bits_per_limb as u32) + BigUint::from(if &input_gagdet.num_of_additions_over_normal_form % BigUint::from(2usize).pow(params.bits_per_limb as u32) != BigUint::from(0usize) { 1usize } else {0usize});
+
+        let result = Self{
+            limbs,
+            num_add_over_normal_form,
+            simulation_phantom: PhantomData,
+        };
+
+        debug_assert!(result.check(), "convert to mul result gagdet from non native gadget: check() failed on output gadget");
+
+        result
+    }
+
+    /// convert `self` to NonNativeFieldGadget. This functions assumes that `self` has the first
+    /// `params.num_limbs` set to 0, otherwise it is pointless to convert them to
+    /// a NonNativeFieldGadget.
```

Comment:
Wouldn't it be a good idea to move the code of pseudo-Mersenne reduction (i.e. the "collapsing") into exactly this function?
So, we would not need a separate `mul_without_prereduce` for the pseudo-Mersenne case, just separate security bounds. As a consequence less code duplication.

Reply:
Not sure the effort would be worth, the duplicated code besides the two functions does not seem that much (just the two `for` loops that perform the pairwise multiplication of the limbs), given that the security bounds are different. Also considering that this part will be heavily changed by refactoring.